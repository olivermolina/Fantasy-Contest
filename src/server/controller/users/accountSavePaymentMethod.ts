import { t } from '~/server/trpc';
import { TRPCError } from '@trpc/server';
import { Session } from '@prisma/client';
import * as yup from '~/utils/yup';
import GIDX, {
  BillingAddressInterface,
  GidxPaymentMethodInterface,
} from '~/lib/tsevo-gidx/GIDX';
import { prisma } from '~/server/prisma';
import { ActionType } from '~/constants/ActionType';

const accountSavePaymentMethod = t.procedure
  .input(
    yup.object({
      fullName: yup.string().required(),
      billingAddress: yup.mixed<BillingAddressInterface>().required(),
      paymentMethod: yup.mixed<GidxPaymentMethodInterface>().required(),
      session: yup.mixed<Session>().required(),
    }),
  )
  .mutation(async ({ input, ctx }) => {
    const userId = ctx.session.user?.id;
    const user = await prisma.user.findUnique({
      where: {
        id: userId,
      },
    });
    if (!userId || !user) {
      throw new TRPCError({
        code: 'INTERNAL_SERVER_ERROR',
        message: 'User not found',
      });
    }

    try {
      const { billingAddress, paymentMethod, session, fullName } = input;

      const gidx = await new GIDX(
        user,
        ActionType.SAVE_PAYMENT_METHOD,
        session,
      );
      return await gidx.savePaymentMethod({
        fullName,
        paymentMethod,
        billingAddress,
      });
    } catch (e: any) {
      throw new TRPCError({
        code: 'INTERNAL_SERVER_ERROR',
        message: e.message,
      });
    }
  });

export default accountSavePaymentMethod;
