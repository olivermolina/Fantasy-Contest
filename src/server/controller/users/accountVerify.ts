import { t } from '~/server/trpc';
import { TRPCError } from '@trpc/server';
import * as yup from '~/utils/yup';
import GIDX, { CustomerProfileResponse } from '~/lib/tsevo-gidx/GIDX';
import { prisma } from '~/server/prisma';
import { Session } from '@prisma/client';
import { ActionType } from '~/constants/ActionType';

const accountVerify = t.procedure
  .input(
    yup.object({
      session: yup.mixed<Session>(),
    }),
  )
  .mutation(async ({ input, ctx }) => {
    const userId = ctx.session.user?.id;
    const user = await prisma.user.findUnique({
      where: {
        id: userId,
      },
    });

    if (!userId || !user) {
      throw new TRPCError({
        code: 'INTERNAL_SERVER_ERROR',
        message: 'User not found',
      });
    }

    const { session } = input;

    if (!session) {
      throw new TRPCError({
        code: 'INTERNAL_SERVER_ERROR',
        message: 'Invalid session!',
      });
    }

    try {
      const gidx = await new GIDX(user, ActionType.GET_CUSTOMER, session);
      const customerProfile: CustomerProfileResponse =
        await gidx.getCustomerProfile();
      if (
        !customerProfile ||
        !customerProfile.ReasonCodes.includes('ID-VERIFIED')
      ) {
        return Promise.reject(Error('Could not identify the customer!'));
      }

      const primaryName = customerProfile.Name.find((name) => name.Primary);
      const primaryAddress = customerProfile.Address.find(
        (address) => address.Primary,
      );

      if (!primaryName || !primaryAddress)
        return Promise.reject(Error('Could not verify the customer details!'));

      return {
        firstname: primaryName.FirstName,
        lastname: primaryName.LastName,
        address1: primaryAddress.AddressLine1,
        address2: primaryAddress.AddressLine2,
        city: primaryAddress.City,
        state: primaryAddress.StateCode,
        postalCode: primaryAddress.PostalCode,
      };
    } catch (e: any) {
      throw new TRPCError({
        code: 'INTERNAL_SERVER_ERROR',
        message: e.message,
      });
    }
  });

export default accountVerify;
