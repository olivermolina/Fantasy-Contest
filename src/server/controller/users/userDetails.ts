import { TRPCError } from '@trpc/server';
import { t } from '~/server/trpc';
import { prisma } from '~/server/prisma';

const userDetails = t.procedure.query(async ({ ctx }) => {
  if (!ctx.session) {
    throw new TRPCError({
      code: 'UNAUTHORIZED',
    });
  }

  const userId = ctx.session.user?.id;
  const user = await prisma.user.findUnique({
    where: {
      id: userId,
    },
  });
  if (!user) {
    throw new TRPCError({
      code: 'INTERNAL_SERVER_ERROR',
      message: 'User not found',
    });
  }
  return user;
});

export default userDetails;
