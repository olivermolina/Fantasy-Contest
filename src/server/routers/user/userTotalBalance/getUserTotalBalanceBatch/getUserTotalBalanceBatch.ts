import { Prisma, User } from '@prisma/client';
import prisma from '~/server/prisma';
import { TRPCError } from '@trpc/server';
import { PaymentStatusCode } from '~/constants/PaymentStatusCode';
import {
  TransactionStatusWithTransaction,
  UserTotalBalanceInterface,
} from '../getUserTotalBalance';
import { generateUserTotalBalancesFromTransactions } from '../generateUserTotalBalancesFromTransactions';

/**
 * This function gets the UserTotalBalance object
 * for a batch of users
 *
 * @param userIds - array of user ids
 * @param trx - transaction client
 */
export const getUserTotalBalanceBatch = async (
  userIds: string[],
  trx?: Prisma.TransactionClient,
): Promise<(UserTotalBalanceInterface & { user: User })[]> => {
  const prismaClient = trx || prisma;
  if (!userIds || userIds.length === 0) {
    return [] as (UserTotalBalanceInterface & { user: User })[];
  }
  const users = await prismaClient.user.findMany({
    where: {
      id: {
        in: userIds,
      },
    },
    include: {
      Transactions: {
        where: {
          TransactionStatuses: {
            some: {
              statusCode: {
                in: [PaymentStatusCode.PENDING, PaymentStatusCode.COMPLETE],
              },
            },
          },
        },
        include: {
          TransactionStatuses: {
            orderBy: {
              created_at: 'asc',
            },
          },
        },
      },
    },
  });
  if (!users) {
    throw new TRPCError({
      code: 'INTERNAL_SERVER_ERROR',
      message: 'User not found',
    });
  }
  const userTotalBalances = await Promise.all(
    users.map(async (user) => {
      const transactionStatuses = user.Transactions.reduce(
        (acc, transaction) => {
          const transactionStatus = transaction.TransactionStatuses.find(
            (transactionStatus) =>
              transactionStatus.statusCode === PaymentStatusCode.COMPLETE,
          );
          if (!transactionStatus) {
            return acc;
          }
          acc.push({
            ...transactionStatus,
            Transaction: transaction,
          });
          return acc;
        },
        [] as TransactionStatusWithTransaction[],
      );

      return {
        ...generateUserTotalBalancesFromTransactions(transactionStatuses),
        user: {
          ...user,
          Transactions: undefined,
        },
      };
    }),
  );
  return userTotalBalances;
};
