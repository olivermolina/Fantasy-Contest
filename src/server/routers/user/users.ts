import { TRPCError } from '@trpc/server';
import { t } from '~/server/trpc';
import { prisma } from '~/server/prisma';
import { Prisma, User, UserType } from '@prisma/client';
import { z } from 'zod';
import type { AgentUsersType } from '~/server/routers/admin/middleware/isAdmin';

// @ts-expect-error Big int was not supported fix here https://github.com/prisma/studio/issues/614#issuecomment-795213237
BigInt.prototype.toJSON = function () {
  return this.toString();
};

/**
 * Set the where filter for the player query
 * @param user - User object
 * @param where - Prisma.UserWhereInput
 */
export const setPlayerWhereFilter = (
  user: User & {
    UserAsAgents: AgentUsersType[];
    SubAdminAgents: AgentUsersType[];
  },
  where: Prisma.UserWhereInput = {},
) => {
  where.type = UserType.PLAYER;
  where.NOT = { id: user.id };
  switch (user.type) {
    case UserType.ADMIN:
      // Do nothing
      break;
    case UserType.SUB_ADMIN:
      // Get all players under the sub admin agents
      where.agentId = {
        in: user.SubAdminAgents.map((agent) => agent.id),
      };
      break;
    case UserType.AGENT:
      // Get users under the agent
      where.agentId = {
        in: user.UserAsAgents.map((agent) => agent.id),
      };
      break;
    default:
  }
};

const users = t.procedure
  .input(
    z
      .object({
        userType: z.nativeEnum(UserType).optional(),
      })
      .optional(),
  )
  .query(async ({ input, ctx }) => {
    if (!ctx.session) {
      throw new TRPCError({
        code: 'UNAUTHORIZED',
      });
    }

    const userId = ctx.session.user?.id;
    const user = await prisma.user.findUnique({
      where: {
        id: userId,
      },
      include: {
        UserAsAgents: {
          include: {
            users: true,
          },
        },
        SubAdminAgents: {
          include: {
            users: true,
          },
        },
      },
    });
    if (!user) {
      throw new TRPCError({
        code: 'INTERNAL_SERVER_ERROR',
        message: 'User not found',
      });
    }

    const where: Prisma.UserWhereInput = {};
    if (!input?.userType || input?.userType === UserType.PLAYER) {
      setPlayerWhereFilter(user, where);
    } else if (input?.userType === UserType.AGENT) {
      where.type = UserType.AGENT;

      switch (user.type) {
        case UserType.ADMIN:
          // Do nothing
          break;
        case UserType.SUB_ADMIN:
          // Get all agents under the sub admin
          where.id = {
            in: user.SubAdminAgents.map((agent) => agent.userId),
          };
          break;
        case UserType.AGENT:
          // return the agent itself
          where.id = userId;
          break;
        default:
          throw new TRPCError({
            code: 'BAD_REQUEST',
            message: `You don't have permission to access the user list`,
          });
      }
    }

    return await prisma.user.findMany({
      where,
      select: {
        id: true,
        email: true,
        username: true,
        referral: true,
        referralCodes: true,
        type: true,
        phone: true,
        DOB: true,
        firstname: true,
        lastname: true,
        address1: true,
        address2: true,
        city: true,
        state: true,
        postalCode: true,
        status: true,
        isFirstDeposit: true,
        created_at: true,
      },
    });
  });

export default users;
