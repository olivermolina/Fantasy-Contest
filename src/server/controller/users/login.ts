import { TRPCError } from '@trpc/server';
import { t } from '~/server/trpc';
import * as yup from '~/utils/yup';
import { prisma } from '~/server/prisma';
import { supabase } from '~/utils/supabaseClient';
import { setAuthResponse } from '~/server/controller/users/setAuthResponse';

const login = t.procedure
  .input(
    yup.object({
      email: yup.string().required(),
      password: yup
        .string()
        .min(8, 'Your password must be at least 8 characters.')
        .required(),
    }),
  )
  .mutation(async ({ ctx, input }) => {
    if (ctx.session.user) {
      throw new TRPCError({
        code: 'BAD_REQUEST',
        message: 'This user is already signed in.',
      });
    }
    const result = await supabase.auth.signIn({
      email: input.email,
      password: input.password,
    });
    if (!result.user) {
      throw new TRPCError({
        code: 'BAD_REQUEST',
        message: 'Invalid email or password!',
      });
    }

    //Check user if exists in prisma
    const prismaUser = await prisma.user.findUnique({
      where: {
        id: result.user.id,
      },
    });

    if (!prismaUser) {
      //Register missing prisma user
      await prisma.user.upsert({
        where: {
          id: result.user.id,
        },
        create: {
          id: result.user.id,
          email: result.user.email!,
          ...result.user.user_metadata,
        },
        update: {
          email: result.user.email!,
          ...result.user.user_metadata,
        },
      });
    }

    setAuthResponse(
      ctx,
      result.session?.access_token,
      result.session?.refresh_token,
    );
  });

export default login;
