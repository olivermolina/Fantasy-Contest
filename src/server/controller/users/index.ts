import accountDeposit from './accountDeposit';
import accountPayout from './accountPayout';
import accountVerify from './accountVerify';
import accountRegister from './accountRegister';
import userDetails from './userDetails';
import checkReferral from './checkReferral';
import login from './login';
import signUp from './signUp';
import createMerchantTransaction from './createMerchantTransaction';
import accountSavePaymentMethod from './accountSavePaymentMethod';

export {
  accountDeposit,
  accountPayout,
  accountRegister,
  accountVerify,
  userDetails,
  checkReferral,
  login,
  signUp,
  createMerchantTransaction,
  accountSavePaymentMethod,
};
