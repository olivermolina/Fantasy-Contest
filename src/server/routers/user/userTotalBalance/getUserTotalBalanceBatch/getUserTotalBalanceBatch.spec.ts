import { getUserTotalBalanceBatch } from './getUserTotalBalanceBatch';
import users from './__mocks__/users.json';

describe('getUserTotalBalanceBatch', () => {
  it('should return user total balance', async () => {
    const prismaClient = {
      user: {
        findMany: jest.fn().mockResolvedValue(users),
      },
    };
    const result = await getUserTotalBalanceBatch(
      [
        '2c1cbb95-5b1f-4bf5-8eda-dc5af334b77b',
        '39b1e5db-d375-4358-aec9-28ce971985de',
      ],
      prismaClient as any,
    );
    expect(result).toMatchSnapshot();
  });
});
