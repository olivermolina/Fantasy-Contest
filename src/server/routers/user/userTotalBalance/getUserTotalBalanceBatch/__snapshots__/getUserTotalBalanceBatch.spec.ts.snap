// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserTotalBalanceBatch should return user total balance 1`] = `
Array [
  Object {
    "creditAmount": 96,
    "totalAmount": 269,
    "totalCashAmount": 173,
    "unPlayedAmount": 964,
    "user": Object {
      "DOB": "1986-11-06T00:00:00.000Z",
      "Transactions": undefined,
      "address1": "",
      "address2": "",
      "agentId": null,
      "city": "",
      "email": "playfree9800@gmail.com",
      "firstname": "ANTRON",
      "id": "2c1cbb95-5b1f-4bf5-8eda-dc5af334b77b",
      "identityStatus": true,
      "isAdmin": false,
      "isFirstDeposit": false,
      "lastname": "HURT",
      "phone": null,
      "postalCode": "",
      "reasonCodes": Array [
        "LL-GEO-US-NJ",
        "LL-GEO-US-NY",
        "LL-BLOCK-PH",
        "LL-ALERT-DIST",
        "LL-BLOCK-US-NY",
        "ID-FAIL",
        "LL-BLOCK",
        "ID-UNKN",
        "ID-VERIFIED",
        "LL-GEO-PH",
        "LL-OUT-US",
      ],
      "referral": null,
      "state": "CA",
      "type": "PLAYER",
      "username": "playfree9800",
    },
    "withdrawableAmount": 173,
  },
  Object {
    "creditAmount": 0,
    "totalAmount": 62.5,
    "totalCashAmount": 62.5,
    "unPlayedAmount": 50,
    "user": Object {
      "DOB": "1997-06-30T00:00:00.000Z",
      "Transactions": undefined,
      "address1": "520 SE 5th ave",
      "address2": "",
      "agentId": null,
      "city": "Fort Lauderdale",
      "email": "sam.kreul@lockspread.com",
      "firstname": "Sam",
      "id": "39b1e5db-d375-4358-aec9-28ce971985de",
      "identityStatus": true,
      "isAdmin": false,
      "isFirstDeposit": false,
      "lastname": "Kreul",
      "phone": "2624228625",
      "postalCode": "53045",
      "reasonCodes": Array [
        "ID-VERIFIED",
        "LL-GEO-US-FL",
      ],
      "referral": "Danielgroup12",
      "state": "FL",
      "type": "PLAYER",
      "username": "Samkreul",
    },
    "withdrawableAmount": 62.5,
  },
]
`;
